pim6d: "show ipv6 pim state" not displaying when OIL is empty

patrasar opened the pull request:
**Problem:**
The cli "show ipv6 pim state" is not displaying when outgoing interface list is empty.
This is fixed now.

**Before Fix:**
`frr# show ipv6 pim state json`
`{`
 ` "ffaa::5":{`
   ` "1100::10":{`
     ` "ens224":{`
  `    },`
    `  "installed":1,`
    `  "isRpt":false,`
    ` "refCount":1,`
    `  "oilListSize":0,`
      `"oilRescan":0,`
      `"lastUsed":0,`
     ` "packetCount":40,`
      `"byteCount":3080,`
    `  "wrongInterface":1`
    `}`
  `}`
`}

`frr# show ipv6 pim state`
`Codes: J -> Pim Join, I -> MLD Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted`
 `Active  Source  Group  RPT  IIF  OIL`

**After fix:**
Case 1:
- "show ipv6 pim state" output with 1 oil.
`frr# show ipv6 pim state`
`Codes: J -> Pim Join, I -> MLD Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted`
 `Active  Source    Group    RPT  IIF     OIL`
` 1       1100::10  ffaa::5  n    ens224`

Case 2:
- "show ipv6 pim state" output with multiple oil.
`frr# show ipv6 pim state`
`Codes: J -> Pim Join, I -> MLD Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted`
`Active  Source    Group    RPT  IIF     OIL`
 `1   1100::10   ffaa::5   n   ens224   ens192( J   )`
                                      ` ens256( J   )`

Case 3:
- "show ipv6 pim state" output with no oil
`frr# show ipv6 pim state`
`Codes: J -> Pim Join, I -> MLD Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted`
 `Active  Source    Group    RPT  IIF     OIL`
 `1       1100::10  ffaa::5  n    ens224`

Closes #13070

donaldsharp commented:
Yeah I have absolutely no idea why we need v4 or v6 defines with specialized code for v4 or v6 in a simple show command.  This seems obtusely and frankly stupidly expensive to maintain.  I think we can find a better way to get the code paths to work better together.

patrasar commented:
> Yeah I have absolutely no idea why we need v4 or v6 defines with specialized code for v4 or v6 in a simple show command. This seems obtusely and frankly stupidly expensive to maintain. I think we can find a better way to get the code paths to work better together.

Agree @donaldsharp. 
The current code is different for V4 and V6. It will be difficult to maintain.
Along with the fix, I added one more commit to make the code common for V4 and V6.

Regards,
Sarita

patrasar commented:
Verified the fix.

frr# show ip pim state
Codes: J -> Pim Join, I -> IGMP Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted
 Active  Source      Group      RPT  IIF     OIL  
 1       110.0.0.10  225.1.1.1  n    ens224       

frr# show ip mroute 
IP Multicast Routing Table
Flags: S - Sparse, C - Connected, P - Pruned
       R - SGRpt Pruned, F - Register flag, T - SPT-bit set
 Source      Group      Flags  Proto  Input   Output  TTL  Uptime    
 110.0.0.10  225.1.1.1  SFP    none   ens224  none    0    --:--:--  

frr# show ipv6 pim state
Codes: J -> Pim Join, I -> MLD Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted
 Active  Source    Group    RPT  IIF     OIL  
 1       1100::10  ffaa::5  n    ens224       

frr# show ipv6 mroute 
IP Multicast Routing Table
Flags: S - Sparse, C - Connected, P - Pruned
       R - SGRpt Pruned, F - Register flag, T - SPT-bit set
 Source    Group    Flags  Proto  Input   Output  TTL  Uptime    
 1100::10  ffaa::5  SFP    none   ens224  none    0    --:--:--

patrasar commented:
This PR will fix 1 part of the issue https://github.com/FRRouting/frr/issues/13070

patrasar commented:
Some random failures.
ci:rerun

mobash-rasool commented:
> Yeah I have absolutely no idea why we need v4 or v6 defines with specialized code for v4 or v6 in a simple show command. This seems obtusely and frankly stupidly expensive to maintain. I think we can find a better way to get the code paths to work better together.

@donaldsharp : This was discussed in detail in the weekly community meeting and then it was decided to keep the v4 and v6 code separate for this CLI because some customer would be using the CLI output and ttable would change the existing format.
Please go through this PR https://github.com/FRRouting/frr/pull/11802 to understand the discussions which happened earlier.

I have asked Sarita to check if she can simplify the code by removing the #ifs if possible.

But this was the only CLI where we could not make the code common due to the problem mentioned in the PR #11802

donaldsharp commented:
I find it extremely hard to believe that we need to do it this way to make the v4 and v6 code work together.   This is a nack from me.  This is a bridge to far.

donaldsharp commented:
```diff --git a/pimd/pim_cmd_common.c b/pimd/pim_cmd_common.c
index 75df09ec35..33760259d6 100644
--- a/pimd/pim_cmd_common.c
+++ b/pimd/pim_cmd_common.c
@@ -1149,12 +1149,18 @@ void pim_show_state(struct pim_instance *pim, struct vty *vty,
                                                    c_oil->cc.wrong_if);
                        }
                }
-#if PIM_IPV == 4
                else
+#if PIM_IPV == 4
                        vty_out(vty, "%-6d %-15pPAs  %-15pPAs  %-3s  %-16s  ",
                                c_oil->installed, oil_origin(c_oil),
                                oil_mcastgrp(c_oil), isRpt ? "y" : "n",
                                in_ifname);
+#else
+                       ttable_add_row(tt, "%d|%pPAs|%pPAs|%s|%s|%s%c",
+                                      c_oil->installed, oil_origin(c_oil),
+                                      oil_mcastgrp(c_oil), isRpt ? "y" : "n",
+                                      in_ifname, out_ifname, ' ');
+
 #endif
 
                for (oif_vif_index = 0; oif_vif_index < MAXVIFS;
```

Would this work?  I think it would

patrasar commented:
> ```diff
> index 75df09ec35..33760259d6 100644
> --- a/pimd/pim_cmd_common.c
> +++ b/pimd/pim_cmd_common.c
> @@ -1149,12 +1149,18 @@ void pim_show_state(struct pim_instance *pim, struct vty *vty,
>                                                     c_oil->cc.wrong_if);
>                         }
>                 }
> -#if PIM_IPV == 4
>                 else
> +#if PIM_IPV == 4
>                         vty_out(vty, "%-6d %-15pPAs  %-15pPAs  %-3s  %-16s  ",
>                                 c_oil->installed, oil_origin(c_oil),
>                                 oil_mcastgrp(c_oil), isRpt ? "y" : "n",
>                                 in_ifname);
> +#else
> +                       ttable_add_row(tt, "%d|%pPAs|%pPAs|%s|%s|%s%c",
> +                                      c_oil->installed, oil_origin(c_oil),
> +                                      oil_mcastgrp(c_oil), isRpt ? "y" : "n",
> +                                      in_ifname, out_ifname, ' ');
> +
>  #endif
>  
>                 for (oif_vif_index = 0; oif_vif_index < MAXVIFS;
> ```
> 
> Would this work? I think it would

Thanks @donaldsharp 
I will check and update on this.
Sorry for the delay in response, as I was working on other deliverables.
Regards,
Sarita

patrasar commented:
> ```diff
> index 75df09ec35..33760259d6 100644
> --- a/pimd/pim_cmd_common.c
> +++ b/pimd/pim_cmd_common.c
> @@ -1149,12 +1149,18 @@ void pim_show_state(struct pim_instance *pim, struct vty *vty,
>                                                     c_oil->cc.wrong_if);
>                         }
>                 }
> -#if PIM_IPV == 4
>                 else
> +#if PIM_IPV == 4
>                         vty_out(vty, "%-6d %-15pPAs  %-15pPAs  %-3s  %-16s  ",
>                                 c_oil->installed, oil_origin(c_oil),
>                                 oil_mcastgrp(c_oil), isRpt ? "y" : "n",
>                                 in_ifname);
> +#else
> +                       ttable_add_row(tt, "%d|%pPAs|%pPAs|%s|%s|%s%c",
> +                                      c_oil->installed, oil_origin(c_oil),
> +                                      oil_mcastgrp(c_oil), isRpt ? "y" : "n",
> +                                      in_ifname, out_ifname, ' ');
> +
>  #endif
>  
>                 for (oif_vif_index = 0; oif_vif_index < MAXVIFS;
> ```
> 
> Would this work? I think it would

Hi @donaldsharp,
I have addressed your comments.
Kindly check.

patrasar commented:
Verified the fix in the below setup

`LHR----RP---FHR`
        `---------`

LHR
`frr# show ipv6 pim state`
`Codes: J -> Pim Join, I -> MLD Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted`
 `Active  Source    Group    RPT  IIF     OIL `            
` 1       *         ff05::2  y    ens224  pim6reg(I    )  `
 `                                        ens192(I    )`   
` 1       *         ffaa::5  y    ens224  pim6reg(I    )  `
`                                         ens192(I    ) `  
 `1       1100::10  ffaa::5  n    ens256  ens192(   * ) `  

RP
`frr# show ipv6 pim state`
`Codes: J -> Pim Join, I -> MLD Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted`
 `Active  Source    Group    RPT  IIF     OIL`            
 `1       *         ff05::2  y    ens192  ens192( J   )`  
 `1       *         ffaa::5  y    ens192  ens192( J   )`  
 `1       1100::10  ffaa::5  n    ens224`                 

FHR
`frr# show ipv6 pim state`
`Codes: J -> Pim Join, I -> MLD Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted`
 `Active  Source    Group    RPT  IIF     OIL`            
` 1       1100::10  ffaa::5  n    ens224  ens256( J   ) `

mobash-rasool commented on pimd/pim_cmd_common.c:
```diff
@@ -1225,6 +1230,7 @@ void pim_show_state(struct pim_instance *pim, struct vty *vty,
 #if PIM_IPV == 4
 					vty_out(vty, "%s%s", out_ifname, flag);
 #else
+					ttable_del_row(tt, tt->nrows - 1);
```
Please add a comment here for future reference, why this has been added.

mobash-rasool commented:
Please fix the below warning:
Report for pim_cmd_common.c | 2 issues
===============================================
< ERROR: else should follow close brace '}'
< #1152: FILE: /tmp/f1-381073/pim_cmd_common.c:1152:

patrasar commented:
> Please fix the below warning: Report for pim_cmd_common.c | 2 issues
> 
> < ERROR: else should follow close brace '}' < #1152: FILE: /tmp/f1-381073/pim_cmd_common.c:1152:

Done

patrasar commented:
ospf failure.
ci:rerun

patrasar commented:
ci:rerun

mobash-rasool commented:
@patrasar : Please rebase the changes, few bgp fixes went in which might fix the CI problem here.

patrasar commented:
> @patrasar : Please rebase the changes, few bgp fixes went in which might fix the CI problem here.

Trying my luck last time :)
